concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -380,9 +382,17 @@ loadSkovPersistentData rp _treeStateDirectory pbsc = do
   (_lastFinalizationRecord, lfStoredBlock) <- liftIO (getLastBlock _db) >>= \case
       Left s -> logExceptionAndThrowTS $ DatabaseInvariantViolation s
       Right r -> return r
-  _lastFinalized <- liftIO (makeBlockPointer lfStoredBlock)
-
-  return SkovPersistentData {
+  -- Check that the block state for the last finalized block is not corrupt. The check works because
+  -- the `BlobRef` to the block state is stored in LMDB only after the block state is stored in the
+  -- blob store. If the reference points to a blob that cannot be read, the blob store can be
+  -- assumed to be corrupted.
+  elfBlob <- liftIO . try $
+    readBlobBSFromHandle (bscBlobStore . PBS.pbscBlobStore $ pbsc) (sbState lfStoredBlock)
```

Comment:
This check is not enough though.

```
concordium-node: too few bytes
From:	Epoch blocks
	demandInput

 :: "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DLE\NUL\NUL\NUL\NUL\NUL\NUL\EOT\DLE\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DLE\SOH\NUL\NUL\NUL\NUL\NUL\NUL\NUL\ETX\NUL\NUL\NUL\NUL\NUL\SOH\141^\NUL\NUL\NUL\NUL\NUL\NULH\136\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DLE\NUL\NUL\NUL\NUL\NUL\NUL]\199\NUL\NUL\NUL\NUL\NUL\NUL\NULe\208\NUL\NUL\EM\238:\tw\a\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\SOH\136\&2O^\NUL\NUL\NUL\NUL\196\EM'\167\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NULe\224\NUL\NUL\NUL\NUL\NUL\NUL\156U\NUL\NUL\NUL\NUL\NUL\NUL\184\144\NUL\NUL\NUL\NUL\NUL\SOH/<\NUL\NUL\NUL\NUL\NUL\SOH/\\\NUL\NUL\NUL\NUL\NUL\NUL\SO\DLE\NUL\NUL\NUL\NUL\NUL\NUL\NUL\141\169\DC4\149\b\169D\vx\156\253\fM\161\180)\187\163W#\202\224\154\ETX\206 \215N\168'L\b\238\&4\165\153D\208\255\t\DC4\\g\139\215\151\174\252\134\179H\183\STX\132\132\217\248\131]\233\US\DELm\194\217\NUL\NUL\NUL\NUL\NUL\NUL\192\184\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\STX\SOH\NUL\NUL\NUL\NUL\NUL\SOH\248\222\NUL\NUL\NUL\NUL\NUL\NUL\252o\NUL\NUL\NUL\NUL\NUL\NULT%;\168sv\172\155eX\187m5\146\217\139\160\198\b\154\&1\128\141\184K\135\&3\222k\191\ACK\233U\139\ETX\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\GS\191\140\243\&8\246tX\225;*\169\128N\233\131\200\NULC\185\USp\NULEy\201\SO\236\208I%\143;\168sv\172\155eX\187m5\146\217\139\160\198\b\154\&1\128\141\184K\135\&3\222k\191\ACK\233U\139\NUL\NUL\NUL\NUL\NUL\SOH,C\NUL\NUL\NUL\NUL\NUL\SOH,\211\NUL\NUL\NUL\NUL\NUL\SOH\141"
CallStack (from HasCallStack):
  error, called at src/Concordium/GlobalState/Persistent/BlobStore.hs:440:19 in concordium-consensus-0.1.0.0-5lPy2W7VWNUGfv8rYBMwgy:Concordium.GlobalState.Persistent.BlobStore
  loadRef, called at src/Concordium/GlobalState/Persistent/BlobStore.hs:619:29 in concordium-consensus-0.1.0.0-5lPy2W7VWNUGfv8rYBMwgy:Concordium.GlobalState.Persistent.BlobStore
```
if I remove just the last byte from the file.

The reason for this is that what is stored at this "root blobref" is the Persistent.BlockPointers, i.e., this
```
data BlockStatePointers (pv :: ProtocolVersion) = BlockStatePointers {
    bspAccounts :: !(Accounts.Accounts pv),
    bspInstances :: !(Instances.Instances pv),
    bspModules :: !(HashedBufferedRef Modules.Modules),
    bspBank :: !(Hashed Rewards.BankStatus),
    bspIdentityProviders :: !(HashedBufferedRef IPS.IdentityProviders),
    bspAnonymityRevokers :: !(HashedBufferedRef ARS.AnonymityRevokers),
    bspBirkParameters :: !(PersistentBirkParameters (AccountVersionFor pv)),
    bspCryptographicParameters :: !(HashedBufferedRef CryptographicParameters),
    bspUpdates :: !(BufferedRef (Updates pv)),
    bspReleaseSchedule :: !(BufferedRef (Map.Map AccountAddress Timestamp)),
    -- FIXME: Store transaction outcomes in a way that allows for individual indexing.
    bspTransactionOutcomes :: !Transactions.TransactionOutcomes,
    -- |Details of bakers that baked blocks in the current epoch. This is
    -- used for rewarding bakers at the end of epochs.
    bspRewardDetails :: !(BlockRewardDetails (AccountVersionFor pv))
}
```

So you have to attempt to load that before you know it succeeds. Note that **loading** this does not mean the entire state will be loaded. Pointers will not be followed.

Reply:
Isn't `sbState lfStoredBlock` a `BlobRef` to `BlockStatePointers`? If so, this exactly what I'm loading from the blobstore (as in `BlobRef a → ByteString`.

I see that what's failing here is deserialization (as in `ByteString → a`).

Reply:
I guess the condition to check before reading bytes from the blob store would be `offset + size + 8 <= fromIntegral bhSize`, and not `offset + size <= fromIntegral bhSize`. I didn't account for the size field having a size.

Reply:
Yes, this seems to have been the case. Good catch!